ospfd/ospf_zebra.c

```diff
@@ -255,6 +256,38 @@ static void ospf_zebra_add_nexthop(struct ospf *ospf, struct ospf_path *path,
 	api->nexthop_num++;
 }
 
+static void ospf_zebra_append_opaque_attr(struct ospf_route *or,
+					  struct zapi_route *api)
+{
+	struct ospf_zebra_opaque ospf_opaque = {};
+
+	/* OSPF path type */
+	snprintf(ospf_opaque.path_type, sizeof(ospf_opaque.path_type), "%s",
+		 ospf_path_type_name(or->path_type));
+
+	switch (or->path_type) {
+	case OSPF_PATH_INTRA_AREA:
+	case OSPF_PATH_INTER_AREA:
+		/* OSPF area ID */
+		(void)inet_ntop(AF_INET, &or->u.std.area_id,
+				ospf_opaque.area_id,
+				sizeof(ospf_opaque.area_id));
+		break;
+	case OSPF_PATH_TYPE1_EXTERNAL:
+	case OSPF_PATH_TYPE2_EXTERNAL:
+		/* OSPF route tag */
+		snprintf(ospf_opaque.tag, sizeof(ospf_opaque.tag), "%u",
+			 or->u.ext.tag);
+		break;
+	default:
+		break;
+	}
+
+	SET_FLAG(api->message, ZAPI_MESSAGE_OPAQUE);
+	api->opaque.length = sizeof(struct ospf_zebra_opaque);
```

Comment:
Can we have here a safety check like with BGP?

```
MIN(sizeof(struct ospf_zebra_opaque),
					ZAPI_MESSAGE_OPAQUE_LENGTH);
```

Reply:
That check isn't necessary anymore since we have this in _lib/route_opaque.h_:
```
static_assert(sizeof(struct bgp_zebra_opaque) <= ZAPI_MESSAGE_OPAQUE_LENGTH,
              "BGP opaque data shouldn't be larger than zebra's buffer");
static_assert(sizeof(struct ospf_zebra_opaque) <= ZAPI_MESSAGE_OPAQUE_LENGTH,
              "OSPF opaque data shouldn't be larger than zebra's buffer");
```